src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputer.java

```diff
@@ -17,33 +17,82 @@
 package com.google.jenkins.plugins.computeengine;
 
 import com.google.api.services.compute.model.Instance;
+import hudson.model.Executor;
+import hudson.model.Result;
+import hudson.model.TaskListener;
+import hudson.remoting.Channel;
 import hudson.slaves.AbstractCloudComputer;
 import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
 import java.util.logging.Level;
-import java.util.logging.Logger;
+import jenkins.model.CauseOfInterruption;
+import lombok.extern.java.Log;
 import org.kohsuke.stapler.DataBoundSetter;
 import org.kohsuke.stapler.HttpRedirect;
 import org.kohsuke.stapler.HttpResponse;
 
+@Log
 public class ComputeEngineComputer extends AbstractCloudComputer<ComputeEngineInstance> {
 
   private volatile Instance instance;
-
-  private static final Logger LOGGER = Logger.getLogger(ComputeEngineCloud.class.getName());
+  private Future<Boolean> preemptedFuture;
 
   public ComputeEngineComputer(ComputeEngineInstance slave) {
     super(slave);
   }
 
-  @Override
-  public ComputeEngineInstance getNode() {
-    return (ComputeEngineInstance) super.getNode();
-  }
-
-  public void onConnected() {
+  void onConnected(TaskListener listener) throws IOException {
     ComputeEngineInstance node = getNode();
     if (node != null) {
       node.onConnected();
+      if (getPreemptible()) {
+        String nodeName = node.getNodeName();
+        log.log(
+            Level.INFO, "Instance " + nodeName + " is preemptive, setting up preemption listener");
+        preemptedFuture = getChannel().callAsync(new PreemptedCheckCallable(listener));
+        getChannel()
+            .addListener(
+                new Channel.Listener() {
+                  @Override
+                  public void onClosed(Channel channel, IOException cause) {
+                    log.log(Level.FINE, "Got channel close event");
+                    if (getPreempted()) {
+                      log.log(
+                          Level.FINE, "Preempted node channel closed, terminating all executors");
+                      getExecutors().forEach(executor -> interruptExecutor(executor, nodeName));
+                    }
+                  }
+                });
+      }
+    }
+  }
+
+  private void interruptExecutor(Executor executor, String nodeName) {
+    log.log(Level.INFO, "Terminating executor " + executor + " node " + nodeName);
+    executor.interrupt(
+        Result.FAILURE,
+        new CauseOfInterruption() {
+          @Override
+          public String getShortDescription() {
+            return "Instance " + nodeName + " was preempted";
+          }
+        });
+  }
+
+  public boolean getPreemptible() {
+    try {
+      return getInstance().getScheduling().getPreemptible();
+    } catch (IOException | NullPointerException e) {
+      return false;
```

Comment:
Swallowing these exceptions seems like a bad idea. What's the expected behavior here?

Reply:
Its bad idea to use checked exceptions everywhere..
I have to catch theme here.. even tho I dont want to do anything with them..
I surly dont want to pass them further..
If they would be runtime exception then I should not care..

Any idea how to do it better?

Reply:
Swallowing the NPE is a red flag for sure, as this could obfuscate other problems. I'd suggest removing that type from the catch block and instead handling it directly in the body.
For the IOException I'd suggest logging it with a descriptive message followed by surfacing the error to allow the call-site to decide what to do with it.

Reply:
I cannot pass it.. cause I will have to add it to signature and its spreading even further..
And with java 8 its hard as hell to work with checked exceptions..

Reply:
I will deal with NPE

Reply:
Now its ok?

Reply:
I think adding a logging statement containing the exception so that the message and stack trace is visible along with returning false would be sufficient in this case, rather than propagating this exception from here.